Python client for btldr

jack-greenberg opened the pull request.

ayushchakra commented on projects/btldr/py_client/btldr.py:
```diff
@@ -0,0 +1,245 @@
+        request_response = self._receive_request_response(ecu_id, timeout)
+
+        if request_response:
+            if request_response["error_code"] != "OK":
```
Are there other specific values in request_response["error_code"] that could be useful to include in the raised exception to help while debugging?

jack-greenberg replied:
Nope, it's just OK or ERROR I think.

jack-greenberg commented on bazel/tools/can_flash.py:
This is a wrapper that can be invoked with Bazel so that you can use it to flash boards over can with their latest firmware:

```sh
$ bazel run --config=16m1 -c opt //vehicle/mkvi/software/air_control:can_flash -- -d can0
```

ayushchakra commented on projects/btldr/ecus.bzl:
```diff
@@ -11,6 +13,9 @@ ECUS = {
     "brakelight_bspd": {
         "btldr_id": "0x718",
     },
+    "brakes": {
```
What's the distinction between `brakelight_bspd` and `brakes`?

jack-greenberg replied:
brakelight_bspd is the name from mkv, brakes is the name from mkvi. It's a thing with how the btldr finds it's ID using the name of the cc_firmware build target. It's not the best way to go about it but it worked and I don't think I'm going to spend much time finding a better solution for now haha.

ayushchakra replied:
fair enough

ayushchakra commented on projects/btldr/py_client/README.md:
```diff
@@ -0,0 +1,70 @@
+is.
+
+<!-- TODO -->
+[Coda]: https://coda.io
```
Update with more specific link?

ayushchakra commented on projects/btldr/py_client/README.md:
```diff
@@ -0,0 +1,70 @@
+has an ECU_ID of 0x700, the `air_control_reset_response` message has a CAN ID of
+`0x700 + 5 = 0x705`.
+
+By default, when you create the BtldrDatabase, if keeps the eight message IDs as
```
nit: it keeps, not if keeps

ayushchakra commented on projects/btldr/py_client/btldr.py:
```diff
@@ -0,0 +1,341 @@
+        # Must be initialized separately
+        self.canbus = None
+
+    """
```
nit: Python docstrings should be below the function signature

ayushchakra commented on projects/btldr/py_client/btldr.py:
```diff
@@ -0,0 +1,341 @@
+        # immediately over and over again until we get a response, but for the
+        # sake of time and completeness, I decided to hardcode a delay.
+        # Hopefully someone will improve upon this one day.
+        time.sleep(1)  # Wait one second after reset before attempting to ping
```
Why can't we just immediately start pinging after the software reset and just loop it until it returns a response?

jack-greenberg replied:
We could, but I think I'm going to leave it for now. It would just be a slight performance improvement and, while it would be more idiomatic, I'd rather just leave it as is for now and focus on other things.

ayushchakra commented on projects/btldr/py_client/btldr.py:
```diff
@@ -0,0 +1,341 @@
+        if ping_resp["current_image"] != "Updater":
+            logging.critical("Failed to place device in updater")
+            exit(1)
+            # raise Exception("Failed to place device in updater")
```
nit: Could delete the commented out exceptions since the logging covers their purpose

ayushchakra commented on projects/btldr/py_client/btldr_database.py:
```diff
@@ -0,0 +1,250 @@
+        query = CANMessage(
+            frame_id=self.ecu_id + 0,
+            name="btldr_query",
+            length=8,
```
nit: abstract out the message lengths and start bits into variables to make them easier to change if we ever want to change the messages

ayushchakra commented on projects/btldr/py_client/cli.py:
```diff
@@ -0,0 +1,124 @@
+
+"""
+This is just a helpful function that takes a time delta (bascailly just a number
+of milliseconds (or is it seconds?)), subtracts it from the current time,
```
Seems like it's seconds, but should clarify the comment

ayushchakra commented on projects/btldr/py_client/tests/test_btldr_client.py:
```diff
@@ -0,0 +1,65 @@
+from formula.projects.btldr.py_client.btldr_database import BtldrDatabase
+
+
+@pytest.fixture()
+def dbc():
+    return BtldrDatabase()
+
+
+@pytest.fixture()
+def ecu_id():
+    return 0x700
```
Seems like these can be removed since they're in `conftest.py`

ayushchakra commented on projects/btldr/py_client/tests/test_btldr_hw.py:
```diff
@@ -0,0 +1,30 @@
+    check.is_not_none(response)
+
+
+def test_flash(canbus, ecu_id):
```
Definitely not necessary for this PR, but would be nice to add a github action that runs any pytest/unit tests to ensure there's no regressions (I might look into this at some point).

jack-greenberg replied:
Yeah, if you wanted to look into this in the future, that'd be awesome! There are lots of opportunities for this.

ayushchakra reviewed (commented):
Overall, looks great! Just one last nit that I noticed across the files, it would be great to add typing to each of your functions just for clarity. Like `def ping_resp(ecu_id, dbc)` would be `def ping_resp(ecu_id: int, dbc: BtldrDatabase) -> CANMessage`

jack-greenberg commented:
Thanks for the review! I implemented all the actionable feedback except where I commented above. I added types to all the functions in `btldr.py` as I think that is where it is most important.
